File: lintcode/BFS/137_clone_graph.py
"""
class UndirectedGraphNode:
     def __init__(self, x):
         self.label = x
         self.neighbors = []
"""

class Solution:
    """
    @param node: A undirected graph node
    @return: A undirected graph node
    """
    def cloneGraph(self, node):
        # write your code here
        new = None
        from queue import Queue

        q = Queue()
        h = set()
        mp = {}
        if not node:
            return new

        q.put(node)
        # h.add(node)
        mp[node.label] = UndirectedGraphNode(node.label)
        new = mp[node.label]
        while not q.empty():
            n = q.get()
            for n1 in n.neighbors:
                if n1.label in mp:
                    continue
                mp[n1.label] = UndirectedGraphNode(n1.label)
                q.put(n1)

        q.put(node)
        h.add(node.label)
        while not q.empty():
            n = q.get()
            c = mp[n.label]
            for n1 in n.neighbors:
                c.neighbors.append(mp[n1.label])
                if n1.label not in h:
                    q.put(n1)
                    h.add(n1.label)

        return new

File: lintcode/BFS/137_clone_graph.go
package main

import "fmt"

/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Neighbors []*Node
 * }
 */

func cloneGraph(node *Node) *Node {
	if node == nil{
		return nil
	}
	visited := make(map[int] bool)
	mp := make(map[int] *Node)
	q := []*Node{node}
	mp[q[0].Val] = &Node{q[0].Val, []*Node{}}

	for;len(q) > 0;{
		t := q[0]
		q = q[1:]
		_, ok := visited[t.Val]
		if !ok {
			for i := 0; i < len(t.Neighbors); i++ {
				if _, val := mp[t.Neighbors[i].Val]; !val {
					mp[t.Neighbors[i].Val] = &Node{t.Neighbors[i].Val, []*Node{}}
				}
				if _, val := visited[t.Neighbors[i].Val]; !val {
					q = append(q, t.Neighbors[i])
				}
				mp[t.Val].Neighbors = append(mp[t.Val].Neighbors, mp[t.Neighbors[i].Val])
			}
		}
		visited[t.Val] = true
	}
	return mp[node.Val]
}